mesonbuild/mesonlib/vsenv.py

```diff
@@ -0,0 +1,100 @@
+import os
+import subprocess
+import json
+import pathlib
+import shutil
+import tempfile
+
+from .. import mlog
+from .universal import MesonException, is_windows
+
+
+bat_template = '''@ECHO OFF
+
+call "{}"
+
+ECHO {}
+SET
+'''
+
+# If on Windows and VS is installed but not set up in the environment,
+# set it to be runnable. In this way Meson can be directly invoked
+# from any shell, VS Code etc.
+def _setup_vsenv(force: bool) -> bool:
+    if not is_windows():
+        return False
+    if os.environ.get('OSTYPE') == 'cygwin':
+        return False
+    if 'Visual Studio' in os.environ['PATH']:
+        return False
+    # VSINSTALL is set when running setvars from a Visual Studio installation
+    # Tested with Visual Studio 2012 and 2017
+    if 'VSINSTALLDIR' in os.environ:
+        return False
+    # Check explicitly for cl when on Windows
+    if shutil.which('cl.exe'):
+        return False
+    if not force:
+        if shutil.which('cc'):
+            return False
+        if shutil.which('gcc'):
+            return False
+        if shutil.which('clang'):
+            return False
+        if shutil.which('clang-cl'):
+            return False
+
```

Comment:
are these `which`es still necessary by the time we get this far?

Reply:
Yes because we still try to setup VS environment even when --vsenv is not specified, but in that case we don't want to activate it when other compilers are available in your PATH.